### DISCUSSION TITLE
When to use function in defining stimulus with timelineVariable (question from workshop example)

### DISCUSSION THREAD

**Discussion by CarolynBuck-Gengler** [jspsych/jsPsych#870] [Q&A]
I am re-watching the workshop from today and working through the example code along with Becky.
At about minute 29+, she is talking about defining the probe stimulus with the anonymous function. I didn't catch why she is doing that instead of simply defining it as 

            stimulus: jsPsych.timelineVariable('probe'),

unless it has to put it into the function to be able to add the surrounding html to add formatting? 
I also looked at the documentation for jsPsych.timelineVariable, where it shows two examples, one called "Standard use as a parameter for a trial" and the other, with the function, for "Invoking immediately in a function". Again, the only difference seems to be adding in additional formatting with html and css. (Which does seem to be necessary in the case in the documentation.)
I guess I just want to understand if there is more here behind this choice than just the formatting. 

Thanks for the great workshop, both of you and the rest of the team! Very well put together.

Carolyn

**Comment by vijaymarupudi** [comment]
That's because `jsPsych.timelineVariable('probe')` returns a function that returns the stimulus. If you'd like to modify the value of the timelineVariable, you can't just use it like a number or a string (`+` or `*` for example), since it's a function.

You can't just put `true` as the second argument then, because when you're defining the trial, jsPsych has no idea what your variables will be.

In that case, using it in a function will help. Since it's a function, jsPsych will not run the stimulus until it's the trial's turn. During the trial's turn, the function is run.

In that point of time, jsPsych does know what variable to provide. However a second parameter is needed to signal to jsPsych that it knows the variable to provide (there are plans to remove the need for the second argument). Then the call will return a string or number, which can be customized as you need.

**Reply by CarolynBuck-Gengler** [reply]
So basically it's because it's being embedded in html markup? Because if no markup is needed (the main difference seems to be adding the surrounding html) the example from the workshop today works without it. I guess I still don't understand the difference between the two examples other than that. (I'm not asking about the true second argument, but about embedding jsPsych.timelineVariable('probe') in a function() {...}, e.g., why

            stimulus: function() {
                return '<span style="font-size:40px;">' + jsPsych.timelineVariable('probe', true) + '</span>';
            },

instead of 

            stimulus: jsPsych.timelineVariable('probe'),

I ran it with the latter, and it ran fine. The documentation also does not really make clear what the difference is.)
Thanks,
Carolyn

**Reply by vijaymarupudi** [reply]
> So basically it's because it's being embedded in html markup?

Yep, that's why!
